pyvisa_sim/testsuite/test_common.py

```diff
@@ -0,0 +1,40 @@
+from typing import List, Optional
+
+import pytest
+
+from pyvisa_sim import common
+
+
+@pytest.mark.parametrize(
+    "data, mask, send_end, want",
+    [
+        (b"1234", None, False, b"1234"),
+        (b"41234", 3, False, b"40004"),
+        # Can't figure this one out. Getting ValueError: bytes must in in range(0, 256)
+        # If I knew more about the purpose of `iter_bytes` then maybe I could
+        # reconcile it, but I don't have time to investigate right now.
+        pytest.param(
+            b"1234",
+            3,
+            True,
+            b"1234",
+            marks=pytest.mark.xfail(
+                reason=(
+                    "ValueError bytes must be in range(0, 256). TODO: figure"
+                    " out correct 'want' value."
+                )
+            ),
+        ),
```

Comment:
Thanks for uncovering a bug !

iter_bytes is meant to clip values to the right number of bits (since serial may use from 5 to 8 bits). When send_end is True the highest data bits should only be set when the message is over. The problem here is that  ~ gives us a signed binary number. So basically the function is bugged (and it is annoying since it is duplicated in pyvisa-py). 

To make it less error prone it is we could refactor it to take the number of data bits and send_end. That way we could write out the mask directly. Can you make the change or do you prefer me to take care of it ?